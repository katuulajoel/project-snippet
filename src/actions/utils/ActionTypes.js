export const STATUS_APPROVED = "approved";
export const STATUS_PENDING = "pending";

export const DOC_TYPE_OTHER = "other";

// Dashboard action types
export const FETCH_NOTIFICATIONS_START = "FETCH_NOTIFICATIONS_START";
export const FETCH_NOTIFICATIONS_SUCCESS = "FETCH_NOTIFICATIONS_SUCCESS";
export const FETCH_NOTIFICATIONS_FAILED = "FETCH_NOTIFICATIONS_FAILED";
export const DELETE_NOTIFICATIONS_START = "DELETE_NOTIFICATIONS_START";
export const DELETE_NOTIFICATIONS_SUCCESS = "DELETE_NOTIFICATIONS_SUCCESS";
export const DELETE_NOTIFICATIONS_FAILED = "DELETE_NOTIFICATIONS_FAILED";
export const CREATE_NOTIFICATION_LOG_START = "CREATE_NOTIFICATION_LOG_START";
export const CREATE_NOTIFICATION_LOG_SUCCESS =
  "CREATE_NOTIFICATION_LOG_SUCCESS";
export const CREATE_NOTIFICATION_LOG_FAILED = "CREATE_NOTIFICATION_LOG_FAILED";

// Invoices action types
export const CREATE_INVOICE_START = "CREATE_INVOICE_START";
export const CREATE_INVOICE_SUCCESS = "CREATE_INVOICE_SUCCESS";
export const CREATE_INVOICE_FAILED = "CREATE_INVOICE_FAILED";
export const CREATE_INVOICE_BATCH_START = "CREATE_INVOICE_BATCH_START";
export const CREATE_INVOICE_BATCH_SUCCESS = "CREATE_INVOICE_BATCH_SUCCESS";
export const CREATE_INVOICE_BATCH_FAILED = "CREATE_INVOICE_BATCH_FAILED";
export const LIST_INVOICES_START = "LIST_INVOICES_START";
export const LIST_INVOICES_SUCCESS = "LIST_INVOICES_SUCCESS";
export const LIST_INVOICES_FAILED = "LIST_INVOICES_FAILED";
export const SEARCH_INVOICES_START = "SEARCH_INVOICES_START";
export const SEARCH_INVOICES_SUCCESS = "SEARCH_INVOICES_SUCCESS";
export const SEARCH_INVOICES_FAILED = "SEARCH_INVOICES_FAILED";
export const RETRIEVE_INVOICE_START = "RETRIEVE_INVOICE_START";
export const RETRIEVE_INVOICE_SUCCESS = "RETRIEVE_INVOICE_SUCCESS";
export const RETRIEVE_INVOICE_FAILED = "RETRIEVE_INVOICE_FAILED";
export const UPDATE_INVOICE_START = "UPDATE_INVOICE_START";
export const UPDATE_INVOICE_SUCCESS = "UPDATE_INVOICE_SUCCESS";
export const UPDATE_INVOICE_FAILED = "UPDATE_INVOICE_FAILED";
export const LIST_MORE_INVOICES_START = "LIST_MORE_INVOICES_START";
export const LIST_MORE_INVOICES_SUCCESS = "LIST_MORE_INVOICES_SUCCESS";
export const LIST_MORE_INVOICES_FAILED = "LIST_MORE_INVOICES_FAILED";
export const SEARCH_MORE_INVOICES_START = "SEARCH_MORE_INVOICES_START";
export const SEARCH_MORE_INVOICES_SUCCESS = "SEARCH_MORE_INVOICES_SUCCESS";
export const SEARCH_MORE_INVOICES_FAILED = "SEARCH_MORE_INVOICES_FAILED";
export const DELETE_INVOICE_START = "DELETE_INVOICE_START";
export const DELETE_INVOICE_SUCCESS = "DELETE_INVOICE_SUCCESS";
export const DELETE_INVOICE_FAILED = "DELETE_INVOICE_FAILED";
export const PAY_INVOICE_START = "PAY_INVOICE_START";
export const PAY_INVOICE_SUCCESS = "PAY_INVOICE_SUCCESS";
export const PAY_INVOICE_FAILED = "PAY_INVOICE_FAILED";
export const ARCHIVE_INVOICE_SUCCESS = "ARCHIVE_INVOICE_SUCCESS";
export const ARCHIVE_INVOICE_START = "ARCHIVE_INVOICE_START";
export const ARCHIVE_INVOICE_FAILED = "ARCHIVE_INVOICE_FAILED";
export const GENERATE_INVOICE_SUCCESS = "GENERATE_INVOICE_SUCCESS";
export const GENERATE_INVOICE_START = "GENERATE_INVOICE_START";
export const GENERATE_INVOICE_FAILED = "GENERATE_INVOICE_FAILED";
export const BULK_ACTION_START = "BULK_ACTION_START";
export const BULK_ACTION_FAILED = "BULK_ACTION_FAILED";
export const BULK_ACTION_SUCCESS = "BULK_ACTION_SUCCESS";
export const DOWNLOAD_INVOICE_CSV_STARTED = "DOWNLOAD_INVOICE_CSV_STARTED";
export const DOWNLOAD_INVOICE_CSV_FAILED = "DOWNLOAD_INVOICE_CSV_FAILED";
export const DOWNLOAD_INVOICE_CSV_SUCCESS = "DOWNLOAD_INVOICE_CSV_SUCCESS";
export const INVOICE_SUMMARY_START = "INVOICE_SUMMARY_START";
export const INVOICE_SUMMARY_SUCCESS = "INVOICE_SUMMARY_SUCCESS";
export const INVOICE_SUMMARY_FAILED = "INVOICE_SUMMARY_FAILED";
export const SEARCH_PROJECTS_START = "SEARCH_PROJECTS_START";
export const SEARCH_PROJECTS_SUCCESS = "SEARCH_PROJECTS_SUCCESS";
export const SEARCH_PROJECTS_FAILED = "SEARCH_PROJECTS_FAILED";
export const SEARCH_MORE_PROJECTS_START = "SEARCH_MORE_PROJECTS_START";
export const SEARCH_MORE_PROJECTS_SUCCESS = "SEARCH_MORE_PROJECTS_SUCCESS";
export const SEARCH_MORE_PROJECTS_FAILED = "SEARCH_MORE_PROJECTS_FAILED";

// Projects Actions types
export const FETCH_PROJECT_START = "FETCH_PROJECT_START";
export const FETCH_PROJECT_SUCCESS = "FETCH_PROJECT_SUCCESS";
export const FETCH_PROJECT_FAILED = "FETCH_PROJECT_FAILED";
// Invites action types
export const GET_PENDING_INVITES_START = "GET_PENDING_INVITES_START";
export const GET_PENDING_INVITES_SUCCESS = "GET_PENDING_INVITES_SUCCESS";
export const GET_PENDING_INVITES_FAILED = "GET_PENDING_INVITES_FAILED";
export const GET_MORE_PENDING_INVITES_START = "GET_MORE_PENDING_INVITES_START";
export const GET_MORE_PENDING_INVITES_SUCCESS =
  "GET_MORE_PENDING_INVITES_SUCCESS";
export const GET_MORE_PENDING_INVITES_FAILED =
  "GET_MORE_PENDING_INVITES_FAILED";
export const INVITE_START = "INVITE_START";
export const INVITE_SUCCESS = "INVITE_SUCCESS";
export const INVITE_FAILED = "INVITE_FAILED";
export const DELETE_INVITE_START = "DELETE_INVITE_START";
export const DELETE_INVITE_SUCCESS = "DELETE_INVITE_SUCCESS";
export const DELETE_INVITE_FAILED = "DELETE_INVITE_FAILED";
