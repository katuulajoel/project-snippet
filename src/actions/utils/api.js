/* eslint-disable prettier/prettier */
import axios from "axios";

axios.defaults.xsrfCookieName = "csrftoken";
axios.defaults.xsrfHeaderName = "X-CSRFToken";
axios.defaults.withCredentials = true;
axios.interceptors.response.use(undefined, (err) => {
  const url = `${
    process.env.REACT_APP_NODE_ENV === "production" ? window.location.origin + "/" : process.env.REACT_APP_BACKEND_ROOT
  }api/auth/verify/`;

  if (
    err.request.responseURL !== url &&
    (err.response.status === 403 || err.response.data.detail === "Authentication credentials were not provided.")
  ) {
    window.location.pathname = `/login`;
  }
  return Promise.reject(err);
});

let BACKEND_PATH = process.env.REACT_APP_BACKEND_ROOT;
let API_PATH = "api/";

export const API_ROOT = `${BACKEND_PATH}${API_PATH}`;

const getEndpointUrl = (path) => {
  return API_ROOT + path;
};

// Auth
export const ENDPOINT_LOGIN = getEndpointUrl("auth/login/");
export const ENDPOINT_LOGOUT = getEndpointUrl("auth/logout/");
export const ENDPOINT_VERIFY = getEndpointUrl("auth/verify/");
export const ENDPOINT_REGISTER = getEndpointUrl("auth/register/");
export const ENDPOINT_EMAIL_VISITOR = getEndpointUrl("auth/visitor/");
export const ENDPOINT_APPLY = getEndpointUrl("apply/");
export const ENDPOINT_INVITE = getEndpointUrl("invite/");
export const ENDPOINT_CHANGE_PASSWORD = getEndpointUrl("auth/password/change/");
export const ENDPOINT_RESET_PASSWORD = getEndpointUrl("auth/password/reset/");
export const ENDPOINT_RESET_PASSWORD_CONFIRM = getEndpointUrl("auth/password/reset/confirm/");

export const USER_TYPE_DEVELOPER = 1;
export const USER_TYPE_PROJECT_OWNER = 2;
export const USER_TYPE_PROJECT_MANAGER = 3;
export const USER_TYPE_DESIGNER = 4;

export const PROJECT_TYPE_WEB = "web";
export const PROJECT_TYPE_MOBILE = "mobile";
export const PROJECT_TYPE_OTHER = "other";

export const PROJECT_DURATION_2_WEEKS = "2w";
export const PROJECT_DURATION_6_MONTHS = "6m";
export const PROJECT_DURATION_PERMANENT = "permanent";

export const STATUS_INITIAL = "initial";
export const STATUS_INTERESTED = "interested";
export const STATUS_UNINTERESTED = "uninterested";

export const INVOICE_TYPE_FINAL = "final";
export const INVOICE_TYPE_SALE = "sale";
export const INVOICE_TYPE_PURCHASE = "purchase";
export const INVOICE_TYPE_CREDIT_NOTE = "credit_nota";
export const INVOICE_TYPES = {
  sale: "Payment",
  purchase: "Payout",
  credit_nota: "Credit Note",
};

// Account
export const ENDPOINT_PROFILE = getEndpointUrl("me/profile/");
export const ENDPOINT_COMPANY = getEndpointUrl("me/company/");
export const ENDPOINT_NOTIFICATIONS = getEndpointUrl("me/notification/");
export const ENDPOINT_ACCOUNT_INFO = getEndpointUrl("me/account/");
export const ENDPOINT_ACCOUNT_SETTINGS = getEndpointUrl("me/settings/");
export const ENDPOINT_USER_INFO = getEndpointUrl("me/user/");
export const ENDPOINT_USER_EDUCATION = getEndpointUrl("me/education/");
export const ENDPOINT_USER_PROJECT = getEndpointUrl("me/project/");
export const ENDPOINT_USER_WORK = getEndpointUrl("me/work/");
export const ENDPOINT_PAYONEER_SIGNUP = getEndpointUrl("payoneer/");

export const ENDPOINT_PROJECTS = getEndpointUrl("projects/");
export const ENDPOINT_USERS = getEndpointUrl("users/");
export const ENDPOINT_DOCUMENTS = getEndpointUrl("documents/");
export const ENDPOINT_PARTICIPATION = getEndpointUrl("participation/");
export const ENDPOINT_PROGRESS_EVENTS = getEndpointUrl("progress-events/");
export const ENDPOINT_PROGRESS_REPORTS = getEndpointUrl("progress-reports/");
export const ENDPOINT_SKILLS = getEndpointUrl("skills/");
export const ENDPOINT_ACTIVITIES = getEndpointUrl("activity/");
export const ENDPOINT_COMMENTS = getEndpointUrl("comments/");
export const ENDPOINT_CHANNELS = getEndpointUrl("channels/");
export const ENDPOINT_MESSAGES = getEndpointUrl("messages/");
export const ENDPOINT_UPLOADS = getEndpointUrl("uploads/");
export const ENDPOINT_INVOICES = getEndpointUrl("invoices/");
export const ENDPOINT_INVOICES_SUMMARY = getEndpointUrl("invoices/summary/");
export const ENDPOINT_INTEREST_POLLS = getEndpointUrl("interest-polls/");

export const ENDPOINT_COUNTRIES = getEndpointUrl("countries/");

export const ENDPOINT_CONTACT_REQUEST = getEndpointUrl("contact-request/");
export const ENDPOINT_INVITE_REQUEST = getEndpointUrl("invite-request/");
export const ENDPOINT_MEDIUM = getEndpointUrl("medium/");
export const ENDPOINT_OEMBED = getEndpointUrl("oembed/");
export const ENDPOINT_SKILL_PAGES = getEndpointUrl("skill-pages/");
export const ENDPOINT_BLOGS = getEndpointUrl("blogs/");
export const ENDPOINT_MIGRATE = getEndpointUrl("migrate/");
export const ENDPOINT_NOTIFICATION_LOG = getEndpointUrl("notification-log/");
export const ENDPOINT_LOG_SEARCH = getEndpointUrl("log/search/");
export const ENDPOINT_VISITORS = getEndpointUrl("visitors/");
export const ENDPOINT_DEVELOPER_RATING = getEndpointUrl("developer-rating/");
export const ENDPOINT_GENERAL_RATING = getEndpointUrl("progress-reports/");

export const ENDPOINT_TEST_RESULTS = getEndpointUrl("results/");

export const flattenJson = (jsonData, key) => {
  let flattenedData = {};

  if (jsonData !== null && jsonData !== undefined) {
    if (jsonData instanceof File) {
      if (key) {
        let flattenedUpdate = {};
        flattenedUpdate[key] = jsonData;
        flattenedData = { ...flattenedData, ...flattenedUpdate };
      }
    } else if (Array.isArray(jsonData)) {
      if (key && jsonData.length) {
        jsonData.forEach((item, idx) => {
          flattenedData = {
            ...flattenedData,
            ...flattenJson(item, `${key}[${idx}]`),
          };
        });
      }
    } else if (typeof jsonData === "object") {
      Object.keys(jsonData).forEach((nestedKey) => {
        flattenedData = {
          ...flattenedData,
          ...flattenJson(jsonData[nestedKey], `${key ? `${key}${key.endsWith("]") ? "" : "."}` : ""}${nestedKey}`),
        };
      });
    } else if (key) {
      let flattenedUpdate = {};
      flattenedUpdate[key] = jsonData;
      flattenedData = { ...flattenedData, ...flattenedUpdate };
    }
  }
  return flattenedData;
};

export const composeFormData = (jsonData) => {
  let flattenedData = flattenJson(jsonData);
  let formData = new FormData();
  Object.keys(flattenedData).forEach((key) => {
    formData.append(key, flattenedData[key]);
  });
  return formData;
};
