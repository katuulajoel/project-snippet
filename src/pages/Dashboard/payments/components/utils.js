import {
  isPayAdmin,
  isClient,
  isPayAdminOrPM,
} from "../../../../components/utils/auth";

export const GENERATE_INVOICE_ACTION = "GENERATE_INVOICE_ACTION";
export const PAY_ACTION = "PAY_ACTION";
export const EDIT_ACTION = "EDIT_ACTION";
export const DELETE_ACTION = "DELETE_ACTION";
export const MARK_AS_PAID_ACTION = "MARK_AS_PAID_ACTION";
export const ARCHIVE_ACTION = "ARCHIVE_ACTION";

export const showAction = (action, invoice) => {
  switch (action) {
    case GENERATE_INVOICE_ACTION:
      return (
        isPayAdmin() &&
        !invoice.finalized &&
        !invoice.last_sent_at &&
        !invoice.paid
      );
    case PAY_ACTION:
      return isClient() && invoice.finalized && !invoice.paid;
    case EDIT_ACTION:
    case DELETE_ACTION:
      return (
        isPayAdminOrPM() &&
        !invoice.paid &&
        !invoice.finalized &&
        !invoice.last_sent_at
      );
    case MARK_AS_PAID_ACTION:
      return (
        isPayAdmin() &&
        !invoice.paid &&
        (invoice.finalized || invoice.last_sent_at)
      );
    case ARCHIVE_ACTION:
      return isPayAdmin() && !invoice.paid;
    default:
      return false;
  }
};

// eslint-disable-next-line no-unused-vars
export const performAction = (action, invoice, props) => {
  switch (action) {
    case GENERATE_INVOICE_ACTION: {
      // onGenerateInvoice(invoice.id, props);
      break;
    }
    case PAY_ACTION: {
      /* openModal(<MakePaymentModal invoice={invoice} />, "Make Payment", true, {
        className: "modal-report modal-make-payment",
      }); */
      break;
    }
    case EDIT_ACTION: {
      // onUpdateInvoice(invoice, props);
      break;
    }
    case DELETE_ACTION: {
      // onDeleteInvoice(invoice.id, props);
      break;
    }
    case MARK_AS_PAID_ACTION: {
      // onMarkPaid(invoice.id, props);
      break;
    }
    case ARCHIVE_ACTION: {
      // onMarkArchived(invoice.id, props);
      break;
    }
  }
};
