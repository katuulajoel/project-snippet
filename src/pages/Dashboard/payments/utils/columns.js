// import { isPayAdmin } from "../../../../components/utils/auth";

//TODO: check if use is pay admin to active batch actions, add ...isPayAdmin() && on line 6

export const getTableColumns = (filter) => [
  ...(!(filter === "archived" || filter === "paid")
    ? [
        {
          Header: " ",
          accessor: "batch_action",
        },
      ]
    : []),
  {
    Header: "Date Created",
    accessor: "created_at",
  },
  {
    Header: "Client / Project / Payment Title",
    accessor: "title",
  },
  {
    Header: "Invoice No.",
    accessor: "invoice",
  },
  {
    Header: "Amount",
    accessor: "amount",
  },
  ...(filter !== "paid"
    ? [
        {
          Header: "Due Date",
          accessor: "due_date",
        },
        {
          Header: "Status",
          accessor: "status",
        },
        ...(filter !== "archived"
          ? [
              {
                Header: "",
                accessor: "actions",
              },
            ]
          : []),
      ]
    : []),
];
