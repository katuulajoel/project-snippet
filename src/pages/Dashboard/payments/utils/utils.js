import {
  isPayAdmin,
  isClient,
  isPayAdminOrPM,
} from "../../../../components/utils/auth";
import * as actions from "./constant"; //TODO: Load just actions needed
import {
  approvePayout,
  generateInvoice,
  markAsArchived,
  markAsPaid,
} from "./paymentActions";

function sumInvoiceAmounts(invoices) {
  return invoices
    .map((invoice) => {
      return invoice.amount || 0;
    })
    .reduce((total, number) => {
      return total + Math.round(number * 100) / 100;
    }, 0);
}

/**
 * Group invoices with the same reference into one batch object
 * @param {*} invoices list of invoices
 * @returns returns array of batch objects
 */
export function batchInvoices(invoices) {
  let batchedInvoices = {};
  invoices?.forEach((invoice) => {
    batchedInvoices[invoice.batch_ref] = [
      ...(batchedInvoices[invoice.batch_ref] || []),
      invoice,
    ];
  });

  let batchObjects = [];
  Object.keys(batchedInvoices).map((ref) => {
    let batchObj = {
      ...batchedInvoices[ref][0],
      invoices: batchedInvoices[ref],
    };
    batchObj.ref = ref;
    batchObj.amount = sumInvoiceAmounts(batchedInvoices[ref]);
    batchObjects.push(batchObj);
  });
  return batchObjects;
}

/**
 * Determine whether to show action or not depending on logged in User
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} invoice invoice if payments action or invoices if payouts action
 * @returns {boolean} false | true
 */
export const showAction = (action, invoice) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION:
      return (
        isPayAdmin() &&
        !invoice.finalized &&
        !invoice.last_sent_at &&
        !invoice.paid
      );
    case actions.PAY_ACTION:
      return isClient() && invoice.finalized && !invoice.paid;
    case actions.EDIT_ACTION:
    case actions.DELETE_ACTION:
      return (
        isPayAdminOrPM() &&
        !invoice.paid &&
        !invoice.finalized &&
        !invoice.last_sent_at
      );
    case actions.MARK_AS_PAID_ACTION:
      return (
        isPayAdmin() &&
        !invoice.paid &&
        (invoice.finalized || invoice.last_sent_at)
      );
    case actions.APPROVE_BATCH_ACTION:
      return isPayAdmin() && invoice.status !== "approved";
    case actions.ARCHIVE_ACTION:
      return isPayAdmin() && !invoice.paid;
    case actions.EDIT_ACTION_BATCH:
    case actions.DELETE_ACTION_BATCH:
      return (
        isPayAdmin() &&
        !invoice.project.archived &&
        !invoice.status !== "approved" &&
        !invoice.paid
      );
  }
};

/**
 * Determine what function to call depending on the action that been triggered
 * @param {string} action constants used for the different actions in payments and payouts
 * @param {*} data invoice if payments action or invoices if payouts action
 */
export const performAction = (action, data) => {
  switch (action) {
    case actions.GENERATE_INVOICE_ACTION: {
      generateInvoice(data.id);
      break;
    }
    case actions.MARK_AS_PAID_ACTION: {
      markAsPaid(data.id);
      break;
    }
    case actions.ARCHIVE_ACTION: {
      markAsArchived(data.id);
      break;
    }
    case actions.APPROVE_BATCH_ACTION: {
      approvePayout(data);
      break;
    }
  }
};

/**
 * filters to be used as query params
 * @param {*} filter : selected payment status;
 * @returns object keys value pairs
 */
export const getPaymentsFilters = (filter) => {
  switch (filter) {
    case "paid":
      return { paid: "True" };
    case "pending":
      return { paid: "False", overdue: "False" };
    case "overdue":
      return { overdue: "True", paid: "False" };
    case "archived":
      return { archived: "True" };
    default:
      return {};
  }
};
