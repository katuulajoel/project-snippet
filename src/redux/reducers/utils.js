export const getIds = (items) => {
  return Array.from(
    new Set(
      items.map((item) => {
        return item.id;
      })
    )
  );
};

export const reduceUser = (state, user, profile, company) => {
  state = state || {};
  user = user || {};
  profile = { ...state.profile, ...user.profile, ...profile };
  company = { ...state.company, ...user.company, ...company };
  [
    ["name", "company"],
    ["bio", "company_bio"],
    ["website", "website"],
    ["vat_number", "vat_number"],
    ["reg_no", "company_reg_no"],
    ["ref_no", "reference_number"],
    ["skills", "skills"],
    ["country", "country"],
    ["city", "city"],
    ["street", "street"],
    ["plot_number", "plot_number"],
    ["postal_code", "postal_code"],
    ["postal_address", "postal_address"],
    ["tel_number", "phone_number"],
  ].forEach((item) => {
    let companyField = item[0],
      profileField = item[1];
    company[companyField] = company[companyField] || profile[profileField];
  });
  return { ...state, ...user, profile, company };
};
