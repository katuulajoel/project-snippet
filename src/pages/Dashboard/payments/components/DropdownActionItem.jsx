/* -------------------------------------------------------------------------- */
/*                            External dependencies                           */
/* -------------------------------------------------------------------------- */
import PropTypes from "prop-types";
import React from "react";
import { DropdownItem } from "reactstrap";

import { showAction, performAction } from "../utils/utils";

const DropdownActionItem = (props) => {
  const { action, invoice, children } = props;
  return showAction(action, invoice) ? (
    <DropdownItem
      className="dropdown-item"
      onClick={() => {
        performAction(action, invoice, props);
      }}
    >
      {children}
    </DropdownItem>
  ) : null;
};

DropdownActionItem.propTypes = {
  action: PropTypes.any, // TODO: add the acceptable prop types
  children: PropTypes.any,
  invoice: PropTypes.any,
};

export default DropdownActionItem;
