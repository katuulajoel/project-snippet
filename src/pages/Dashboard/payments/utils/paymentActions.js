import React from "react";
import {
  archiveInvoice,
  generateInvoice as generateInvoiceAction,
  updateInvoice,
} from "../../../../actions/InvoiceActions";
import ModalHeader from "../../../../components/ModalHeader";
import { openConfirm } from "../../../../components/utils/modals";
import store from "../../../../store";

export function generateInvoice(invoiceId) {
  openConfirm(
    <div>Are you sure you want to generate an invoice for this payment?</div>,
    null,
    true,
    { ok: "Yes" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Generate Invoice" }}
    />
  ).then(() => {
    store.dispatch(generateInvoiceAction(invoiceId));
  });
}

export function markAsPaid(invoiceId) {
  openConfirm(
    <div>Are you sure you want to mark this invoice as paid?</div>,
    null,
    true,
    { ok: "Yes" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Mark as Paid" }}
    />
  ).then(() => {
    store.dispatch(updateInvoice(invoiceId, { paid: true }));
  });
}

export function markAsArchived(invoiceId) {
  openConfirm(
    <div>Are you sure you want to archive this invoice?</div>,
    null,
    true,
    { ok: "Yes" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Archive invoice" }}
    />
  ).then(() => {
    store.dispatch(archiveInvoice(invoiceId));
  });
}

export function approvePayout(invoices) {
  openConfirm(
    <div>Are you sure you want to approve this payout?</div>,
    null,
    true,
    { ok: "Approve", cancel: "Cancel" },
    <ModalHeader
      style={{ paddingBottom: "8px" }}
      options={{ title: "Approve payout" }}
    />
  ).then(() => {
    invoices.forEach((invoice) => {
      store.dispatch(updateInvoice(invoice.id, { status: "approved" }));
    });
  });
}
