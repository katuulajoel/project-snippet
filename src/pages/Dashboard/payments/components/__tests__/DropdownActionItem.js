/* eslint-disable no-unused-vars */
import React from "react";
import DropdownActionItem from "../DropdownActionItem";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import * as utils from "../../utils/utils";
import { mount } from "enzyme/build";

const invoice = {
  id: "42",
  project: {
    id: 1,
    user: {},
    owner: {
      display_name: "Client Tunga",
    },
    title: "Test Project (updated)",
  },
  full_title: "Test Project (updated): Pay Dami",
  subtotal: "1000.00",
  total_amount: "1000.00",
  download_url: "https://work.tunga.io/api/invoices/42/download/?format=pdf",
  due_at: "2021-06-09T23:59:59.999999",
  is_overdue: true,
  title: "Pay Dami",
  type: "purchase",
  amount: "1000.00",
  number: "2021/3/P1/42",
  status: "pending",
  paid: false,
  finalized: true,
  archived: false,
  batch_ref: "7234db4d-24b9-4705-b4dc-e65f20b32e6e",
  created_at: "2021-05-26T18:05:21.072944",
};

describe("ActionItem tests", () => {
  it("Should match snapshot test", () => {
    jest.spyOn(utils, "showAction").mockReturnValueOnce(true);
    const tree = renderer
      .create(
        <BrowserRouter>
          <DropdownActionItem />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
