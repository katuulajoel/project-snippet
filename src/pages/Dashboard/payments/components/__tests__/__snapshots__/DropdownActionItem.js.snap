// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionItem tests Should match snapshot test 1`] = `
<button
  className="dropdown-item dropdown-item"
  onClick={[Function]}
  role="menuitem"
  tabIndex="0"
  type="button"
/>
`;
